### PR TITLE
fix!: correct `camelcase` rule schema for `allow` option

### PR DESCRIPTION
**Pull request by eMerzh**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

The Json schema of the camelCase rule validate that the first element is a string instead of validating that it's an array of string

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Correct the Json Schema

#### Is there anything you'd like reviewers to focus on?

Nope

### CODE REVIEW

**Comment by mdjermanovic** [comment]
Per our [semver policies](https://github.com/eslint/eslint?tab=readme-ov-file#semantic-versioning-policy), this is a breaking change (Part of the public API is removed or changed in an incompatible way > Rule schemas).

@eslint/eslint-tsc Since we've already released v9 rc, looks like we'll have to leave this for ESLint v10?

**Comment by eMerzh** [comment]
@mdjermanovic i understand your position but if i understand, this shouldn't break your CI as normally you could "only" have "one string" in the array... thus this technically allows you do to more 

Also it's only in the validation side, not on the lint itself

**Review by Shinigami92 — APPROVED** [review]
This will fix https://github.com/eslint-types/eslint-define-config/pull/278 semi-automatically (after the generate rule-types script is executed)

**Comment by fasttime** [comment]
This looks like a bug that should be fixed. I'm noting that the current logic allows arbitrary elements in the `allow` array after the first string without any kind of validation. If one of the elements it's a Symbol, or an object that can't be formatted, or even an invalid regex pattern like `"foo)"`, the rule will crash.

```js
/*eslint camelcase: ["error", {allow: ["foo", {toString: null}]}]*/

function bar_baz() {
}
```

[**Playground link**](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgY2FtZWxjYXNlOiBbXCJlcnJvclwiLCB7YWxsb3c6IFtcImZvb1wiLCB7dG9TdHJpbmc6IG51bGx9XX1dKi9cblxuZnVuY3Rpb24gYmFyX2JheigpIHtcbn1cbiIsIm9wdGlvbnMiOnsiZW52Ijp7fSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e30sImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=)

We could look into how to improve option validation for this rule in a non-breaking way, but that wouldn't be the end of the job, so probably it's best to schedule the proper fix for v10 if it can't be done earlier.

**Comment by Tanujkanti4441** [comment]
What i also noticed that if the `allow: [""]` option has an empty string then rule doesn't report the camelcase variables name - [playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgY2FtZWxjYXNlOiBbXCJlcnJvclwiLCB7YWxsb3c6IFtcIlwiXX1dKi9cblxudmFyIGJhcl9iYXogPSBcInN0cmluZ1wiO1xuIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIn19fQ==) and also rule doesn't check whether the string has an underscore or not, should we validate the option this way?

**Comment by nzakas** [comment]
@mdjermanovic If we are categorizing this as a bug fix, then I'm okay with including in the next RC. RCs (and betas) really are designed to catch bugs. As long as we're not adding anything new, I think it's appropriate to include in v9.

**Review by aladdin-add — CHANGES_REQUESTED** [review]
👍 to include it in v9, but we need:

1. add a few tests.
2. add it in the migration guide: https://github.com/eslint/eslint/blob/main/docs/src/use/migrate-to-9.0.0.md

**Comment by mdjermanovic** [comment]
> @mdjermanovic If we are categorizing this as a bug fix, then I'm okay with including in the next RC. RCs (and betas) really are designed to catch bugs. As long as we're not adding anything new, I think it's appropriate to include in v9.

Sounds good to me, marking this as accepted.

**Comment by eMerzh** [comment]
i was thinking about adding those in tests/lib/rules/camelcase.js but it is really the place? 🤔 

also i was thinking of a test with 2-3 allows and 1 or 2 with numbers or objects in options, did you think about something else?

**Comment by mdjermanovic** [comment]
RuleTester currently doesn't support testing invalid options. (it will when we implement https://github.com/eslint/rfcs/tree/main/designs/2023-test-rule-errors).

I think it would be fine to just add a couple more tests with valid options (`allow` with multiple string elements).

File: docs/src/use/migrate-to-9.0.0.md
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:39]:
```suggestion
* [`camelcase` allow option only accepts an array of strings](#camelcase)
```

File: docs/src/use/migrate-to-9.0.0.md
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:440]:
```suggestion
**To address:** If ESLint reports invalid configuration for this rule, update your configuration.
```

File: tests/lib/rules/camelcase.js
Comment (mdjermanovic) [inline tests/lib/rules/camelcase.js:347]: This code would be valid without options, maybe you intended to use some other variable name?

File: docs/src/use/migrate-to-9.0.0.md
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:436]:
```suggestion
**Related issue(s):** [#18042](https://github.com/eslint/eslint/issues/18042)

## <a name="camelcase"></a> `camelcase` allow option only accepts an array of strings
```

Seems that the line for a related issue for the previous item was accidentally deleted.

Comment (eMerzh) [reply docs/src/use/migrate-to-9.0.0.md:436]: arg, shit, fixing

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks! Would like another review before merging.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Since this is a breaking change, it requires another approval from @eslint/eslint-tsc before merging.

**Pull request merged**
